### PR TITLE
don't go through llvm for const indexing

### LINKED ISSUE #3170: Report const index out-of-bound condition earlier
Currently if you try indexing into a const vector, in a const expr, we report the bounds-overrun late in compilation -- during translation.  We should notice it earlier, in the const evaluation pass.

### PR DESCRIPTION
**Pull request by oli-obk**
closes #3170

### CODE REVIEW

File: src/librustc/middle/const_eval.rs [hunk 1]
```diff
@@ -873,6 +899,33 @@ pub fn eval_const_expr_partial<'tcx>(tcx: &ty::ctxt<'tcx>,
...
+        };
+        match arr {
+            const_val::Array(ref v) if idx as usize >= v.len() => signal!(e, IndexOutOfBounds),
+            const_val::Array(v) => try!(eval_const_expr_partial(tcx, &*v[idx as usize], None)),
+            const_val::Repeat(_, n) if idx >= n => signal!(e, IndexOutOfBounds),
+            const_val::Repeat(elem, _) => try!(eval_const_expr_partial(tcx, &*elem, None)),
+            _ => signal!(e, IndexedNonVec),
+        }
+      }
+      ast::ExprVec(ref arr) => const_val::Array(arr.clone()), // FIXME: eval elements?
```
Comment (pnkfelix) [inline src/librustc/middle/const_eval.rs]: ... I would imagine you do need to eval the elements ...

**Comment by oli-obk** [comment]
hmmm... getting quite a few test failures. I guess I broke creating references to const arrays and their elements. I'm evaluating this.

```
    [run-pass] run-pass/check-static-slice.rs
    [run-pass] run-pass/const-autoderef.rs
    [run-pass] run-pass/const-enum-vec-index.rs
    [run-pass] run-pass/const-fields-and-indexing.rs
    [run-pass] run-pass/const-str-ptr.rs
    [run-pass] run-pass/issue-17233.rs
    [run-pass] run-pass/syntax-extension-source-utils.rs
```

File: src/librustc/middle/const_eval.rs (hunk 1 above)
Comment (oli-obk) [inline src/librustc/middle/const_eval.rs]: > But what happens if no one ever indexes the element in a constant expression?

Then they get evaluated in https://github.com/oli-obk/rust/blob/const_indexing/src/librustc_trans/trans/consts.rs#L694-717

Comment (oli-obk) [inline src/librustc/middle/const_eval.rs]: I have updated this to simply memorize the `ast::NodeId` like tuples and structs do. Note: tuples and structs also do lazy evaluation for the field accesses.

**Comment by oli-obk** [comment]
this isn't as trivial as I thought. I'm closing this until I've got a working version. sorry about that.

**Pull request closed without merging**